### PR TITLE
Allow wildcards in test file names

### PR DESCRIPTION
**Pull request by tshastry**
Lets you do things like:

```
python test_discovery.py examples/fruits "test*"
```

### CODE REVIEW

File: test_discovery.py [hunk 1]
```diff
@@ -159,22 +159,27 @@ def test_all(test_file):
...
 
         test_failures.append(entity_results['test_failures'])
         test_failures.append([intent_results.get('test_failures', None)])
-            
+
         output_dict[test_no] = intent_results
-    
+
     test_failures = [i for s in test_failures for i in s]
     test_failures = [_ for _ in test_failures if _]
```
Comment (ageojo) [inline test_discovery.py]: ?

Comment (tshastry) [reply test_discovery.py]: YAPF

File: test_discovery.py [hunk 2]
```diff
@@ -263,6 +268,32 @@ def run_all_tests_in_directory(directory, custom_directory, tests, shutdown):
...
+    Expands any wildcard filenames in the list of tests
+    
+    >>> expand_wildcard_tests('examples/fruits', ['test*'])
+    ['tests.txt', 'tests_negative.txt']
+    
+    >>> expand_wildcard_tests('skip_this_directory', ['no_wildcards'])
+    ['no_wildcards']
+    """
+    if any('*' in t for t in tests):
+        wildcard_tests = [t for t in tests if '*' in t]
```
Comment (ageojo) [inline test_discovery.py]: for loop repetition can be avoided

File: test_discovery.py [hunk 3]
```diff
@@ -263,6 +268,32 @@ def run_all_tests_in_directory(directory, custom_directory, tests, shutdown):
...
+        expanded_filenames = [ \
+            glob.glob(join_path(directory, add_extension_if_missing(t))) \
+            for t in wildcard_tests \
+        ]
+        flattened_filenames = [i for s in expanded_filenames for i in s]
```
Comment (ageojo) [inline test_discovery.py]: variable names in place of `i` or `s`.

File: test_discovery.py [hunk 4]
```diff
@@ -159,22 +159,27 @@ def test_all(test_file):
...
         total_characters += entity_results['total_characters']
 
         test_failures.append(entity_results['test_failures'])
         test_failures.append([intent_results.get('test_failures', None)])
-            
+
         output_dict[test_no] = intent_results
-    
+
     test_failures = [i for s in test_failures for i in s]
```
Comment (ageojo) [inline test_discovery.py]:
```suggestion
    test_failures = [i for s in test_failures for i in s if i]
```
if you added `if i`, could you avoid the second for loop? 
name variables :)

Comment (tshastry) [reply test_discovery.py]: The point of this is to flatten the nested list. Not sure the named variables add anything here.

File: test_discovery.py (hunk 3 above)
Comment (tshastry) [reply test_discovery.py]: Again, this is simply to flatten the nested list. I don’t know what value variable names provide here.

File: test_discovery.py (hunk 4 above)
Comment (tshastry) [reply test_discovery.py]: No, the loop can’t be avoided by putting an if statement. This scheme is simply to flatten the nested lists.

**Pull request merged**